pyfar/dsp/regularization.py

```diff
@@ -0,0 +1,337 @@
+            signal = pf.dsp.pad_zeros(signal, (n_samples-signal.n_samples))
+            self._target = pf.dsp.pad_zeros(self._target,
+                                            (n_samples-self._target.n_samples))
+
+        # Assign signal to class attribute, so further processing methods can
+        # use zero padded signal (e.g. invert())
+        self._signal = signal
+
+        # Call private method to get regularization factors
+        if self._regu_type == "frequency range":
+            regu = self._get_regularization_from_frequency_range(signal)
+        elif self._regu_type == "magnitude spectrum":
+            regu = self._get_regularization_from_regularization_signal(signal)
+
+        # Normalize regularization factors to energy of signal
+        if self._normalize == 'energy':
+            regu.freq *= np.max(np.abs(signal.freq)**2)
+
+        if self._target is not None:
+            regu = self._get_regularization_from_target(signal, regu)
+
+        return regu
+
+    def _get_regularization_from_target(self, signal, regu):
+        """Get regularization using a target function"""
+        # calculate target function from regularization (Norcross 2006)
+        regu_abs_sq = regu.freq * np.conj(regu.freq)
+        sig_abs_sq = signal.freq * np.conj(signal.freq)
+
+        A_eq = 1 / (1 + self._beta * (regu_abs_sq / sig_abs_sq))
+
+        # Apply passed target function
+        A = A_eq * self._target.freq
+
+        return pf.FrequencyData(A, signal.frequencies)
+
+
+    def _get_regularization_from_frequency_range(self, signal):
+        """Get regularization factors from frequency range."""
+
+        if not isinstance(signal, pf.Signal):
+            raise TypeError(f"Regularization of type '{self._regu_type}' "
+                         "requires an input signal of type pyfar.Signal.")
+
+        # arrange regularization factors for inside and outside frequency range
+        regu_inside = np.zeros(signal.n_bins, dtype=np.double)
+        regu_outside = np.ones(signal.n_bins, dtype=np.double)
+
+        # index for crossfade at lower frequency limit
+        idx_xfade_lower = signal.find_nearest_frequency(
+            [self._frequency_range[0]/np.sqrt(2), self._frequency_range[0]])
+
+        regu_final = _cross_fade(regu_outside, regu_inside, idx_xfade_lower)
+
+        # index for crossfade at upper frequency limit
+        if self._frequency_range[1] < signal.sampling_rate/2:
+            idx_xfade_upper = signal.find_nearest_frequency([
+                self._frequency_range[1],
+                np.min([self._frequency_range[1]*np.sqrt(2),
+                        signal.sampling_rate/2])])
+
+            # crossfade regularization factors at frequency limits
+            regu_final = _cross_fade(regu_final, regu_outside, idx_xfade_upper)
+
+        if self._target is None:
+            # control amount of regularization using beta
+            regu_final *= self._beta
```

Comment:
I would suggest to apply beta only during the inversion. It should not be in the returned regularization function.